### PR TITLE
fix(compiler): Allow components to use any style of selector. Fixes #…

### PR DESCRIPTION
**Pull request by jelbourn**
…1602

### CODE REVIEW

File: modules/angular2/src/render/dom/compiler/compiler.ts [hunk 1]
```diff
@@ -49,7 +50,12 @@ export class DomCompiler extends RenderCompiler {
       directives: [directiveMetadata]
     });
     var template = DOM.createTemplate('');
-    DOM.appendChild(DOM.content(template), DOM.createElement(directiveMetadata.selector));
+
+    // In the case of multiple valid selectors, use the first one.
+    var selector = CssSelector.parse(directiveMetadata.selector)[0];
+    var directiveElement = selector.createMatchingElement();
```
Comment (tbosch) [inline modules/angular2/src/render/dom/compiler/compiler.ts]: The selector should output a string, not an actual Element.
The current master also expects a string here (that's why this PR can't be merged any more)

File: modules/angular2/src/render/dom/compiler/selector.ts [hunk 2]
```diff
@@ -91,6 +92,19 @@ export class CssSelector {
     this.element = element;
   }
 
+  createMatchingElement(): HTMLElement {
```
Comment (tbosch) [inline modules/angular2/src/render/dom/compiler/selector.ts]: Please return a string instead of an actual Element here.

Comment (tbosch) [inline modules/angular2/src/render/dom/compiler/selector.ts]: Can you add unit tests for this into `selector_spec.ts`?

File: modules/angular2/test/render/dom/compiler/compiler_common_tests.ts [hunk 3]
```diff
@@ -74,6 +75,28 @@ export function runCompilerCommonTests() {
...
+             id: 'id',
+             selector: 'marquee.jazzy[size=huge]',
+             type: DirectiveMetadata.COMPONENT_TYPE
+           });
+
+           compiler.compileHost(dirMetadata)
+               .then((protoView) => {
+                 let element = DOM.firstChild(
+                     DOM.content(resolveInternalDomProtoView(protoView.render).rootElement));
+                 expect(DOM.tagName(element)).toEqual('MARQUEE');
```
Comment (tbosch) [inline modules/angular2/test/render/dom/compiler/compiler_common_tests.ts]: Test the special cases in `selector_spec.ts`

**Comment by jelbourn** [comment]
Tobias' refactoring of Shadow DOM emulation forced this change to be rewritten. @vsavkin, will you take a look at the new version?

File: modules/angular2/src/render/dom/compiler/compiler.ts [hunk 4]
```diff
@@ -58,10 +59,12 @@ export class DomCompiler extends RenderCompiler {
       directives: [directiveMetadata],
       encapsulation: ViewEncapsulation.NONE
     });
-    return this._compileView(
-        hostViewDef, new TemplateAndStyles(
-                         `<${directiveMetadata.selector}></${directiveMetadata.selector}>`, []),
-        ViewType.HOST);
+
+    let selector = CssSelector.parse(directiveMetadata.selector)[0];
```
Comment (vsavkin) [inline modules/angular2/src/render/dom/compiler/compiler.ts]: Do you know if we handle empty selector?

Comment (jelbourn) [inline modules/angular2/src/render/dom/compiler/compiler.ts]: If you omit selector completely, you get an error
```
EXCEPTION: The selector "undefined" did not match any elements
```

If you use an empty string as a selector, you get
```
EXCEPTION: TypeError: Cannot read property 'componentDirective' of undefined
```

Filed #3464

**Comment by jelbourn** [comment]
Closed via c20a5d65d86c57227a1f3d3cc0b0ca2b085eac48

**Pull request closed without merging**